### PR TITLE
Crash when opening plot options on a sample logs plot

### LINKED ISSUE #27770: Can't open figure options on sample logs plot
### Steps to reproduce the behavior
1. Load deltat_tdc_dolly_1529.bin.md5 in the unit test data.
2. Right click and select Show Sample Logs
3. Scroll down and click Temp_ChannelA.
4. Double-click the plot so it pops out.
5. Click the cog in the toolbar.

### LINKED ISSUE #28635: Script generation does not work for sample log plots
scripts generated from plots extracted from the sample logs do not run 

axes.plot is missing ExperimentInfo=ts_logs.run(), FullTime=False, LogName="C9_SLAVE_FREQUENCY"

Need to test this for "normal" and md (multiple expt info) sample log plots

### PR DESCRIPTION
**Pull request by NickDraper**
**Description of work.**

This fixes two issues where the solutions are intertwined.
Both affect sample log plots when they have been popped out of the sample log dialog into separate plots:
1. Plot options & the right-click menu crash
2. Scripts generated from the plot do not recreate the original graph (they error)

Aspects of the work:
1. Store log plotting details into the artist class so they can be accessed later.
2. When generating scripts extract the log name, Filtered and expt info to populate the generation string.
3. Add protection from None being passed to C++ when trying to determine the ws index for a spectra number (log plots don't have one).
4. Enable recreation of the plot after changes in the plot options dialog.  This required a few limited changes where it was expected that the data should have errors, but doesn't for sample logs.
5. Hide the "Fit" button is the plot contains a sample log plot.

**To test:**

Test the plot options dialog for a sample log plot:

1. Load some data
   ``` python
   ts_logs = Load("ENGINX00228061")
   ```
1. Right-click and select Show Sample Logs
1. Scroll down and click any float series log.
1. Double-click the plot so it pops out.
1. Click the cog in the toolbar.
1. Make changes in the plot options dialog, make sure they are made to the graph.
1. Also check the right-click context menu works:
   1. For sample log plots
   1. For "normal plots"
       1. One issue I found was that a normal plot may be rescaled after changing some aspect of the curve (line the line color).  It looks like it is being rescaled to make room for error bars that are not shown.  However, this is already present on master and is not associated with this PR.

Then test script generation:

1. Open Workbench load a file with some logs ``` ts_logs = Load("ENGINX00228061")```
1. Workspace->show sample logs
1. pick a log with a plot you would recognize (w has a square wave pattern)
1. Double click on the plot to open it externally
1. There should not be a fit button visible at the top of the plot window
1. on the created plot click the Script icon (don't click the cog or right-click as this will crash - that is a different issue)
1. Paste the script into a python window and execute it to check it recreates the same plot.
1. Now check that other plot types that could be scripted still work.

This script also quickly loads and plots a sample log plot
``` python
ts_logs = Load("ENGINX00228061")

fig, axes = plt.subplots(edgecolor='#ffffff', num='Figure 1', subplot_kw={'projection': 'mantid'})
axes.plot(ts_logs, ExperimentInfo=0, Filtered=True, LogName='w', color='#1f77b4', drawstyle='steps-post', label='w')
axes.set_xlabel('Time (s)')
axes.legend().set_draggable(True)

fig.show()
```

Fixes #28635

Fixes #27770

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/mantid/plots/mantidaxes.py [hunk 1]
```diff
@@ -268,9 +268,10 @@ def artists_workspace_has_errors(self, artist):
             if any([workspace.readE(i)[spec_num] != 0 for i in range(0, workspace.getNumberHistograms())]):
                 return True
         else:
-            workspace_index = workspace.getIndexFromSpectrumNumber(spec_num)
-            if any(workspace.readE(workspace_index) != 0):
-                return True
+            if spec_num is not None:
```
Comment (PhilColebrooke) [inline Framework/PythonInterface/mantid/plots/mantidaxes.py:271]: Since this if-statement is the only thing inside the else it can be changed to `elif`.

**Review by PhilColebrooke — CHANGES_REQUESTED** [review]
The options can now be opened, but changing any of the options in the Curves tab produces this error:
```
Traceback (most recent call last):
  File "C:\Users\ilb25526\mantid\qt\python\mantidqt\widgets\plotconfigdialog\presenter.py", line 66, in apply_properties
    tab.apply_properties()
  File "C:\Users\ilb25526\mantid\qt\python\mantidqt\widgets\plotconfigdialog\curvestabwidget\presenter.py", line 71, in apply_properties
    self._replot_selected_curve(plot_kwargs)
  File "C:\Users\ilb25526\mantid\qt\python\mantidqt\widgets\plotconfigdialog\curvestabwidget\presenter.py", line 129, in _replot_selected_curve
    new_curve = FigureErrorsManager.replot_curve(ax, curve, plot_kwargs)
  File "C:\Users\ilb25526\mantid\qt\applications\workbench\workbench\plotting\figureerrorsmanager.py", line 118, in replot_curve
    new_curve = ax.replot_artist(curve, errorbars=True, **plot_kwargs)
  File "C:\Users\ilb25526\mantid\Framework\PythonInterface\mantid\plots\mantidaxes.py", line 379, in replot_artist
    workspace_index = workspace.getIndexFromSpectrumNumber(spec_num)
Boost.Python.ArgumentError: Python argument types in
    None.getIndexFromSpectrumNumber(Workspace2D, NoneType)
did not match C++ signature:
    getIndexFromSpectrumNumber(class Mantid::API::MatrixWorkspace {lvalue} self, int spec_no)
```

**Comment by NickDraper** [comment]
Fixing that requires some of the changes in #28639 I'll take this pr to draft and merge that in and close that PR.

File: docs/source/release/v5.1.0/mantidworkbench.rst [hunk 2]
```diff
@@ -27,13 +27,15 @@ Improvements
 - The axes tab in the figure options can now be used to set the limits, label, and scale of the z-axis on 3D plots.
 - The plot toolbar now shows the correct buttons for 3D plots.
 - On 3D plots you can now double-click on the z-axis to change its limits or label.
+- Plots extracted from "Show Sample Logs" by double clicking the plot can now be converted to a python script, just like other workbnehc plots.
```
Comment (PhilColebrooke) [inline docs/source/release/v5.1.0/mantidworkbench.rst:30]:
```suggestion
- Plots extracted from "Show Sample Logs" by double clicking the plot can now be converted to a python script, just like other Workbench plots.
```

**Review by PhilColebrooke — CHANGES_REQUESTED** [review]
If I open the plot options on a sample logs plot, change something in the Curves tab, click apply, then change something else and click apply again, I get this error:
```
Traceback (most recent call last):
  File "C:\Users\ilb25526\mantid\qt\applications\workbench\workbench\plotting\figureerrorsmanager.py", line 119, in replot_curve
    new_curve = ax.replot_artist(curve, errorbars=True, **plot_kwargs)
  File "C:\Users\ilb25526\mantid\Framework\PythonInterface\mantid\plots\mantidaxes.py", line 394, in replot_artist
    workspace, spec_num = self.get_artists_workspace_and_spec_num(artist)
  File "C:\Users\ilb25526\mantid\Framework\PythonInterface\mantid\plots\mantidaxes.py", line 200, in get_artists_workspace_and_spec_num
    raise ValueError("Artist: '{}' not tracked by axes.".format(artist))
ValueError: Artist: '[<matplotlib.lines.Line2D object at 0x00000155D8340130>]' not tracked by axes.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\ilb25526\mantid\qt\python\mantidqt\widgets\plotconfigdialog\presenter.py", line 66, in apply_properties
    tab.apply_properties()
  File "C:\Users\ilb25526\mantid\qt\python\mantidqt\widgets\plotconfigdialog\curvestabwidget\presenter.py", line 71, in apply_properties
    self._replot_selected_curve(plot_kwargs)
  File "C:\Users\ilb25526\mantid\qt\python\mantidqt\widgets\plotconfigdialog\curvestabwidget\presenter.py", line 129, in _replot_selected_curve
    new_curve = FigureErrorsManager.replot_curve(ax, curve, plot_kwargs)
  File "C:\Users\ilb25526\mantid\qt\applications\workbench\workbench\plotting\figureerrorsmanager.py", line 121, in replot_curve
    new_curve = cls._replot_mpl_curve(ax, curve, plot_kwargs)
  File "C:\Users\ilb25526\mantid\qt\applications\workbench\workbench\plotting\figureerrorsmanager.py", line 136, in _replot_mpl_curve
    remove_curve_from_ax(curve)
  File "C:\Users\ilb25526\mantid\qt\python\mantidqt\widgets\plotconfigdialog\curvestabwidget\__init__.py", line 70, in remove_curve_from_ax
    curve.remove()
TypeError: remove() takes exactly one argument (0 given)
```

**Review by PhilColebrooke — CHANGES_REQUESTED** [review]
I'm still getting the error when I change a setting in the curves tab, click apply, then change something again and click apply.

I also get this error when I click the Fit button on a sample logs plot:
```
Traceback (most recent call last):
  File "C:\Users\ilb25526\mantid\qt\applications\workbench\workbench\plotting\figuremanager.py", line 356, in fit_toggle
    self.fit_browser.show()
  File "C:\Users\ilb25526\mantid\qt\python\mantidqt\widgets\fitpropertybrowser\fitpropertybrowser.py", line 143, in show
    self._add_spectra(allowed_spectra)
  File "C:\Users\ilb25526\mantid\qt\python\mantidqt\widgets\fitpropertybrowser\fitpropertybrowser.py", line 75, in _add_spectra
    self.addAllowedSpectra(name, spec_list)
TypeError: index 0 has type 'NoneType' but 'int' is expected
```

**Comment by NickDraper** [comment]
Pated from slack:

> Phil I can't reproduce the crash on #28636 on my branch, are you sure you pulled in the changes?  Otherwise we need a call so I can see how you crash it.
> 
> 
> 
> 
> 
> 
> Phil Colebrooke:house_with_garden:  9:45 AM
> Call
> Call Ended
> Ended at 9:48 AM - Lasted 3 minutes
> 
> You both joined
> 
> Nick Draper  9:52 AM
> I can reproduce the fit button crash though, so thanks for that.
> :sunglasses:
> 1
> 
> 
> Phil Colebrooke:house_with_garden:  11:35 AM
> So I'm still seeing the error when I just check out your branch without merging master... :thinking_face:
> 
> Nick Draper  12:42 PM
> Could you give me the exact steps you follow to get an error for me to try to reproduce (edited) 
> 
> Phil Colebrooke:house_with_garden:  12:47 PM
> Basically the same as what you were doing on the call. I run the script in your PR to produce a sample logs plot, open the options, change something in the curves tab, click apply, changing something again, click apply
> 12:48
> Maybe we need to get a third person to try to see if there's something wrong with my setup
> 
> Nick Draper  12:48 PM
> what is the top commit message if you type git log?
> 
> Phil Colebrooke:house_with_garden:  12:53 PM
> "Ensure the graph is replotted with stored artist info"
> 
> Nick Draper  12:53 PM
> Hmm, that is the most recent commit.
> 12:54
> Let me solve the fit button issue, and then we will try to get someone else involved

**Comment by ConorMFinn** [comment]
Fails for me too. The second applied change causes an error:

Note: This does not just happen when apply is clicked. Changing something, clicking okay, and then reopening the options and changing something else causes the same error.
```Traceback (most recent call last):
  File "/home/conor/repos/mantid-review/qt/applications/workbench/workbench/plotting/figureerrorsmanager.py", line 119, in replot_curve
    new_curve = ax.replot_artist(curve, errorbars=True, **plot_kwargs)
  File "/home/conor/repos/mantid-review/Framework/PythonInterface/mantid/plots/mantidaxes.py", line 466, in replot_artist
    self.remove_artists_if(lambda art: art == artist)
  File "/home/conor/repos/mantid-review/Framework/PythonInterface/mantid/plots/mantidaxes.py", line 318, in remove_artists_if
    return self._remove_artists_if(unary_predicate)
  File "/home/conor/repos/mantid-review/Framework/PythonInterface/mantid/plots/mantidaxes.py", line 380, in _remove_artists_if
    is_empty = self._remove_artist_info_if(artist_info, unary_predicate)
  File "/home/conor/repos/mantid-review/Framework/PythonInterface/mantid/plots/mantidaxes.py", line 403, in _remove_artist_info_if
    is_empty_list = [workspace_artist.remove_if(self, unary_predicate) for workspace_artist in artist_info]
  File "/home/conor/repos/mantid-review/Framework/PythonInterface/mantid/plots/mantidaxes.py", line 403, in <listcomp>
    is_empty_list = [workspace_artist.remove_if(self, unary_predicate) for workspace_artist in artist_info]
  File "/home/conor/repos/mantid-review/Framework/PythonInterface/mantid/plots/mantidaxes.py", line 1491, in remove_if
    self._remove(axes, artists_to_remove)
  File "/home/conor/repos/mantid-review/Framework/PythonInterface/mantid/plots/mantidaxes.py", line 1499, in _remove
    artist.remove()
  File "/usr/lib/python3/dist-packages/matplotlib/artist.py", line 151, in remove
    self._remove_method(self)
  File "/usr/lib/python3/dist-packages/matplotlib/axes/_base.py", line 1783, in <lambda>
    line._remove_method = lambda h: self.lines.remove(h)
ValueError: list.remove(x): x not in list

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/conor/repos/mantid-review/qt/python/mantidqt/widgets/plotconfigdialog/presenter.py", line 66, in apply_properties
    tab.apply_properties()
  File "/home/conor/repos/mantid-review/qt/python/mantidqt/widgets/plotconfigdialog/curvestabwidget/presenter.py", line 70, in apply_properties
    self._replot_selected_curve(plot_kwargs)
  File "/home/conor/repos/mantid-review/qt/python/mantidqt/widgets/plotconfigdialog/curvestabwidget/presenter.py", line 128, in _replot_selected_curve
    new_curve = FigureErrorsManager.replot_curve(ax, curve, plot_kwargs)
  File "/home/conor/repos/mantid-review/qt/applications/workbench/workbench/plotting/figureerrorsmanager.py", line 121, in replot_curve
    new_curve = cls._replot_mpl_curve(ax, curve, plot_kwargs)
  File "/home/conor/repos/mantid-review/qt/applications/workbench/workbench/plotting/figureerrorsmanager.py", line 136, in _replot_mpl_curve
    remove_curve_from_ax(curve)
  File "/home/conor/repos/mantid-review/qt/python/mantidqt/widgets/plotconfigdialog/curvestabwidget/__init__.py", line 70, in remove_curve_from_ax
    curve.remove()
  File "/usr/lib/python3/dist-packages/matplotlib/artist.py", line 151, in remove
    self._remove_method(self)
  File "/usr/lib/python3/dist-packages/matplotlib/axes/_base.py", line 1783, in <lambda>
    line._remove_method = lambda h: self.lines.remove(h)
ValueError: list.remove(x): x not in list```

**Comment by martyngigg** [comment]
For independent verification I can also reproduce the error after clicking apply twice in the plot options. Does this help reproduce the problem:
sample-logs-plot-options

I tried both checking out the branch and also doing `git test-pr 28636` so that it was merged with master and the result was the same.

File: Framework/PythonInterface/mantid/plots/mantidaxes.py [hunk 3]
```diff
@@ -418,17 +439,39 @@ def replot_artist(self, artist, errorbars=False, **kwargs):
         """
         kwargs['distribution'] = not self.get_artist_normalization_state(artist)
         workspace, spec_num = self.get_artists_workspace_and_spec_num(artist)
-        self.remove_artists_if(lambda art: art == artist)
-        if kwargs.get('axis', None) == MantidAxType.BIN:
-            workspace_index = spec_num
+
+        #check if it is a sample log plot
```
Comment (PhilColebrooke) [inline Framework/PythonInterface/mantid/plots/mantidaxes.py:443]:
```suggestion
        # check if it is a sample log plot
```

File: Framework/PythonInterface/mantid/plots/mantidaxes.py [hunk 4]
```diff
@@ -418,17 +439,39 @@ def replot_artist(self, artist, errorbars=False, **kwargs):
...
+        if spec_num is None:
+            #error bars never make sense for sample log plots
```
Comment (PhilColebrooke) [inline Framework/PythonInterface/mantid/plots/mantidaxes.py:445]: This comment is repeated further down where you set `errorbars` to false.
```suggestion
        if spec_num is None:
```

File: Framework/PythonInterface/mantid/plots/mantidaxes.py [hunk 5]
```diff
@@ -418,17 +439,39 @@ def replot_artist(self, artist, errorbars=False, **kwargs):
...
+            sample_log_plot_details = self.get_artists_sample_log_plot_details(artist)
+            kwargs['LogName'] = sample_log_plot_details[0]
+            if sample_log_plot_details[1] is not None:
+                kwargs['Filtered'] = sample_log_plot_details[1]
+            if sample_log_plot_details[2] is not None:
+                kwargs['ExperimentInfo'] = sample_log_plot_details[2]
+            # error bar plots do not make sense for log plots
+            errorbars=False
```
Comment (PhilColebrooke) [inline Framework/PythonInterface/mantid/plots/mantidaxes.py:453]:
```suggestion
            errorbars = False
```

File: Framework/PythonInterface/mantid/plots/mantidaxes.py [hunk 6]
```diff
@@ -418,17 +439,39 @@ def replot_artist(self, artist, errorbars=False, **kwargs):
...
         if errorbars:
-            new_artist = self.errorbar(workspace, wkspIndex=workspace_index, **kwargs)
+            new_artist = self.errorbar(workspace, **kwargs)
         else:
-            new_artist = self.plot(workspace, wkspIndex=workspace_index, **kwargs)
+            # if it is a sample log plot
+            if spec_num is None:
+                new_artist = self.plot(workspace, **kwargs)[0]
+            else:
+                new_artist = self.plot(workspace, **kwargs)[0]
```
Comment (PhilColebrooke) [inline Framework/PythonInterface/mantid/plots/mantidaxes.py:474]: You have the if and the else both doing the same thing
```suggestion
            new_artist = self.plot(workspace, **kwargs)[0]
```

File: Framework/PythonInterface/mantid/plots/mantidaxes.py [hunk 7]
```diff
@@ -1388,6 +1437,11 @@ def __init__(self,
         :param workspace_name: String. The name of the associated workspace
         :param spec_num: The spectrum number of the spectrum used to plot the artist
         :param is_spec: True if spec_num represents a spectrum rather than a bin
+        :param log_name: string. The name of the plotted log, the
```
Comment (PhilColebrooke) [inline Framework/PythonInterface/mantid/plots/mantidaxes.py:1440]:
```suggestion
        :param log_name: string. The name of the plotted log
```

File: qt/applications/workbench/workbench/plotting/plotscriptgenerator/test/test_plotscriptgeneratorlines.py [hunk 8]
```diff
@@ -89,6 +89,20 @@ def test_generate_plot_command_returns_correct_string_for_line2d(self):
                                                   convert_args_to_string(None, kwargs)))
         self.assertEqual(expected_command, output)
 
+    def test_generate_plot_command_returns_correct_string_for_sample_log(self):
+        kwargs = copy(LINE2D_KWARGS)
+        kwargs["drawstyle"] = 'steps-post'
+        kwargs.update({"LogName":"my_log","ExperimentInfo":0,"Filtered":True})
```
Comment (PhilColebrooke) [inline qt/applications/workbench/workbench/plotting/plotscriptgenerator/test/test_plotscriptgeneratorlines.py:95]:
```suggestion
        kwargs.update({"LogName": "my_log", "ExperimentInfo": 0, "Filtered": True})
```

**Review by PhilColebrooke — DISMISSED** [review]
This works well now 👍 just got a couple of suggestions to tidy up the code

**Review by alicerussell1 — APPROVED** [review]
Looks good to me, works well 👍

**Pull request merged**